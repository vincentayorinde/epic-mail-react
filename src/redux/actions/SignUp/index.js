import { axiosRequest, populateLocalStorage } from '../../../utils';

export const cleanSignUp = () => ({ type: 'CLEAN_SIGNUP' });
export const signUpPending = () => ({ type: 'SIGNUP_PENDING', payload: { isCompleted: false, isLoading: true, error: null } });
export const signUpSuccess = user => ({ type: 'SIGNUP_SUCCESS', payload: { user, isCompleted: true, isAuthenticated: true } });
export const signUpError = payload => ({ type: 'SIGNUP_ERROR', payload });
export const signUpUser = () => ({ type: 'SIGNUP_USER', payload: { isSubmit: true } });
export const setUser = payload => ({ type: 'SET_USER', payload });

export const signUpAction = (signUpData, history) => async (dispatch) => {
  // dispatch(signInPending());
  dispatch(cleanSignUp());
  // dispatch(signUpUser());
  try {
    const result = await axiosRequest({ url: '/api/v2/auth/signup', payload: signUpData, method: 'post' }); 
    populateLocalStorage(result.data);
    dispatch(signUpSuccess(result));
  } catch (e) {
    /* istanbul ignore next */
    const { response } = e;
    /* istanbul ignore next */
    const message = response.data.message || response;
    /* istanbul ignore next */
    dispatch(signUpError(message));
  }
};

export const getUser = () => (dispatch) => {
  /* istanbul ignore next */
  if (localStorage.token && localStorage.user) {
    let user = localStorage.getItem('user');
    user = JSON.parse(user);
    dispatch(setUser(user));
  }
};
