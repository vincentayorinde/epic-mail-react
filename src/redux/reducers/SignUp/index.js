const initialState = {
  isLoading: false,
  error: null,
  user: {},
  isAuthenticated: false,
  isCompleted: false,
  isSubmit: false,
};

const signUpReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SIGNUP_PENDING':
      return {
        ...state,
        ...action.payload,
      };
    case 'SIGNUP_SUCCESS':
      return {
        ...state,
        ...action.payload,
        isSubmit: false,
      };
    case 'SIGNUP_ERROR':
      return {
        ...state,
        error: action.payload,
        isSubmit: false,
      };
    case 'CLEAN_SIGNUP':
      return {
        ...state,
        isCompleted: false,
        error: null,
      };
    case 'SET_USER':
      return {
        ...state,
        isAuthenticated: true,
        user: { ...action.payload },
      };
    case 'SIGNUP_USER':
      return {
        ...state,
        isAuthenticated: true,
        isCompleted: true,
        ...action.payload,
      };
    default:
      return state;
  }
};

export default signUpReducer;
