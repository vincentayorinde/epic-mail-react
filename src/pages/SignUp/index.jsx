import './index.scss';
import React, { Fragment, useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import { signUpAction, cleanSignUp } from '../../redux/actions/SignUp';
import '../../assets/css/react-toastify.scss';
import Input from '../../components/Input';
import Button from '../../components/Button';

const SignUp = (props) => {
  const {
    isCompleted, history, isSubmit, error
  } = props;
  const [values, setValues] = useState({
    firstname: '',
    lastname: '',
    email: '',
    mobile: '',
    password: '',
    confirmPassword: ''
  });
  const onChange = (e) => {
    e.persist();
    setValues(prevState => ({ ...prevState, [e.target.name]: e.target.value }));
  };

  const submit = (e) => {
    e.preventDefault();
    if (values.password !== values.confirmPassword) {
      toast.error(<h4 className="text-center">password must match</h4>);
      return;
    }
    props.onSignUp(values);
  };

  useEffect(() => {
    if (isCompleted) {
      toast.success(<h4 id="toast-success" className="text-center">Registration successful</h4>);
      /* istanbul ignore next */
      history && history.push('/inbox');
    }
    if (error) {
      if (Array.isArray(error)) {
        for (let i = 0; i < error.length; i += 1) {
          toast.error(<h4 className="text-center">{error[i].message}</h4>);
        }
      } else {
        toast.error(<h4 className="text-center">{error}</h4>);
      }
    }
    /* istanbul ignore next */
    return () => {
      /* istanbul ignore next */
      props.cleanUp();
    };
  }, [isCompleted, error]);
  return (
    <Fragment>
    <div className="container-1">
    <div className="box-1">
      <div className="content">
          <p>Experience</p>
          <p>the new way to</p>
          <p> send messages</p>
          <p>Use EPIC Mail <em>today!</em></p>
          <div className="links">
              <ul>
                <li><a href="#">Terms</a></li>
                <li><a href="#">Privacy</a></li>
                <li><a href="#">APIs</a></li>
              </ul>
            </div>
      </div>
      
    </div>
    <div className="box-2">
      <div className="signup">
        <p>Welcome to EPIC Mail</p>
          <h2>Sign Up</h2>
            <form onSubmit={submit}>
              <p>Email</p>
              <Input 
              inputType="email"
              name="email"
              id="email"
              placeholder="e.g. johndoe@examle.com"
              onChange={onChange}
              pattern="^[\w.]+@[\w]{2,20}.[a-z]{2,10}$"
              title="user@email.com" 
              required />
              <p>First Name</p>
              <Input type="text" onChange={onChange} name="firstname" id="firstname" required />
              <p>Last Name</p>
              <Input type="text" onChange={onChange} name="lastname" id="lastname" required />
              <p>Password</p>
              <Input type="password" onChange={onChange} name="password" id="password" required />
              <p>Re-Type Password</p>
              <Input type="password" onChange={onChange} name="confirmPassword" id="confirmPassword" required />
              <p>Mobile</p>
              <Input type="text" onChange={onChange} name="mobile" id="mobile" required placeholder="e.g. 08011122233" />
              <Button
                isSubmit={isSubmit} 
                type="submit" 
                name="Start" 
              />
            </form> 
            <br /><br />
            <i className="fas fa-chevron-right"></i> Already have an account? <a href="sign-in.html">Sign In</a>
      </div>
    </div>
    <div className="box-3">
        <div className="content">
        </div>
    </div>
    </div>
  </Fragment>
  );
};

const mapStateToProps = state => ({
  error: state.auth.error,
  user: state.auth.user,
  isAuthenticated: state.auth.isAuthenticated,
  isSubmit: state.auth.isSubmit,
  isCompleted: state.auth.isCompleted
});


/* istanbul ignore next */
export const onSignUp = newUser => signUpAction(newUser);
export const cleanUp = () => cleanSignUp();

export const SignUpComponent = SignUp;

export default connect(mapStateToProps, { onSignUp, cleanUp })(SignUp);