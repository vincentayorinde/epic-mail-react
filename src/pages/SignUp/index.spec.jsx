import React from 'react';
import { shallow, mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { BrowserRouter as Router } from 'react-router-dom';
import SignUp, { SignUpComponent, cleanUp } from './index';
import store from '../../redux/storeConfig';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Sign Up View Component', () => {
  it('Should render without errors', () => {
    const component = shallow(
      <Router>
        <SignUp store={store} />
      </Router>
    );
    expect(component).toMatchSnapshot();
  });

  it('Should Render With Children Component', () => {
    const component = mount(
      <Router>
        <SignUp store={store} />
      </Router>
    );
    expect(component).toMatchSnapshot();
    expect(component.find('button')).toHaveLength(1);
    expect(component.find('input')).toHaveLength(6);
  });

  it('Should intiate onChange Props', () => {
    const event = {
      preventDefault() { },
      target: { value: 'vincent', name: 'email' }
    };
    const mockOnSignUpFn = jest.fn();
    const component = mount(
      <Router>
        <SignUpComponent onSignUp={mockOnSignUpFn} />
      </Router>
    );
    const inputTag = component.find('input').at(0);
    const submitBtn = component.find('button').at(0);
    inputTag.simulate('change', event);
    submitBtn.simulate('submit',
      { preventDefault() { } });
    expect(mockOnSignUpFn.mock.calls.length).toBe(1);
  });

  test('USE EFFECT => Should set toast if isCompleted is true', () => {
    const signUpStore = mockStore({
      auth:
      {
        error: null,
        user: {},
        isAuthenticated: false,
        isCompleted: true,
        isSubmit: false
      }
    });
    const component = mount(
      <Router>
        <SignUp store={signUpStore} />
      </Router>
    );

    const p = component.find('p').at(5);
    expect(p.text()).toEqual('Email');
  });

  test('Should initiate errors if error exist and is a string', () => {
    const signUpStore = mockStore({
      auth:
      {
        error: 'errors',
        user: {},
        isAuthenticated: false,
        isCompleted: false,
        isSubmit: false
      }
    });
    const component = mount(
      <Router>
        <SignUp store={signUpStore} />
      </Router>
    );

    const p = component.find('p').at(5);
    expect(p.text()).toEqual('Email');
  });

  test('Should initiate errors if error exist and is an array', () => {
    const signUpStore = mockStore({
      auth:
      {
        error: [{ message: 'errors' }],
        user: {},
        isAuthenticated: false,
        isCompleted: false,
        isSubmit: false
      }
    });
    const component = mount(
      <Router>
        <SignUp store={signUpStore} />
      </Router>
    );

    const p = component.find('p').at(5);
    expect(p.text()).toEqual('Email');
  });

  test('Should run clean up action ', () => {
    expect(cleanUp()).toEqual({ type: 'CLEAN_SIGNUP' });
  });

  it('Should check if password match', () => {
    const mockOnSignUpFn = jest.fn();
    const component = mount(
      <Router>
        <SignUpComponent onSignUp={mockOnSignUpFn} />
      </Router>
    );
    const inputTag = component.find('input');
    inputTag.at(0).simulate('change', {
      target:
        { name: 'email', value: 'vincent@epicmail.com' }
    });
    inputTag.at(1).simulate('change', {
      target:
        { name: 'firstName', value: 'vincent' }
    });
    inputTag.at(2).simulate('change', {
      target:
        { name: 'lastName', value: 'ayorinde' }
    });
    inputTag.at(3).simulate('change', {
      target:
        { name: 'password', value: 'password1' }
    });
    inputTag.at(4).simulate('change', {
      target:
        { name: 'confirmPassword', value: 'password2' }
    });
    const submitBtn = component.find('button').at(0);
    submitBtn.simulate('submit',
      { preventDefault() { } });
    expect(mockOnSignUpFn.mock.calls.length).toBe(0);
  });
  
});