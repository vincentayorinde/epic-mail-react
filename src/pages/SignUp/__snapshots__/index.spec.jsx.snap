// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Up View Component Should Render With Children Component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(SignUp)
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <SignUp
        cleanUp={[Function]}
        error={null}
        isAuthenticated={false}
        isCompleted={false}
        isSubmit={false}
        onSignUp={[Function]}
        store={
          Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
        user={Object {}}
      >
        <div
          className="container-1"
        >
          <div
            className="box-1"
          >
            <div
              className="content"
            >
              <p>
                Experience
              </p>
              <p>
                the new way to
              </p>
              <p>
                 send messages
              </p>
              <p>
                Use EPIC Mail 
                <em>
                  today!
                </em>
              </p>
              <div
                className="links"
              >
                <ul>
                  <li>
                    <a
                      href="#"
                    >
                      Terms
                    </a>
                  </li>
                  <li>
                    <a
                      href="#"
                    >
                      Privacy
                    </a>
                  </li>
                  <li>
                    <a
                      href="#"
                    >
                      APIs
                    </a>
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <div
            className="box-2"
          >
            <div
              className="signup"
            >
              <p>
                Welcome to EPIC Mail
              </p>
              <h2>
                Sign Up
              </h2>
              <form
                onSubmit={[Function]}
              >
                <p>
                  Email
                </p>
                <Input
                  id="email"
                  inputType="email"
                  name="email"
                  onChange={[Function]}
                  pattern="^[\\\\w.]+@[\\\\w]{2,20}.[a-z]{2,10}$"
                  placeholder="e.g. johndoe@examle.com"
                  required={true}
                  title="user@email.com"
                >
                  <input
                    id="email"
                    name="email"
                    onChange={[Function]}
                    pattern="^[\\\\w.]+@[\\\\w]{2,20}.[a-z]{2,10}$"
                    placeholder="e.g. johndoe@examle.com"
                    required={true}
                    title="user@email.com"
                    type="email"
                  />
                </Input>
                <p>
                  First Name
                </p>
                <Input
                  id="firstname"
                  name="firstname"
                  onChange={[Function]}
                  required={true}
                  type="text"
                >
                  <input
                    id="firstname"
                    name="firstname"
                    onChange={[Function]}
                    required={true}
                    type="text"
                  />
                </Input>
                <p>
                  Last Name
                </p>
                <Input
                  id="lastname"
                  name="lastname"
                  onChange={[Function]}
                  required={true}
                  type="text"
                >
                  <input
                    id="lastname"
                    name="lastname"
                    onChange={[Function]}
                    required={true}
                    type="text"
                  />
                </Input>
                <p>
                  Password
                </p>
                <Input
                  id="password"
                  name="password"
                  onChange={[Function]}
                  required={true}
                  type="password"
                >
                  <input
                    id="password"
                    name="password"
                    onChange={[Function]}
                    required={true}
                    type="password"
                  />
                </Input>
                <p>
                  Re-Type Password
                </p>
                <Input
                  id="confirmPassword"
                  name="confirmPassword"
                  onChange={[Function]}
                  required={true}
                  type="password"
                >
                  <input
                    id="confirmPassword"
                    name="confirmPassword"
                    onChange={[Function]}
                    required={true}
                    type="password"
                  />
                </Input>
                <p>
                  Mobile
                </p>
                <Input
                  id="mobile"
                  name="mobile"
                  onChange={[Function]}
                  placeholder="e.g. 08011122233"
                  required={true}
                  type="text"
                >
                  <input
                    id="mobile"
                    name="mobile"
                    onChange={[Function]}
                    placeholder="e.g. 08011122233"
                    required={true}
                    type="text"
                  />
                </Input>
                <Button
                  isSubmit={false}
                  name="Start"
                  type="submit"
                >
                  <button
                    type="submit"
                  >
                    Start
                  </button>
                </Button>
              </form>
              <br />
              <br />
              <i
                className="fas fa-chevron-right"
              />
               Already have an account? 
              <Link
                to="/signin"
              >
                <a
                  href="/signin"
                  onClick={[Function]}
                >
                  Sign In
                </a>
              </Link>
            </div>
          </div>
          <div
            className="box-3"
          >
            <div
              className="content"
            />
          </div>
        </div>
      </SignUp>
    </Connect(SignUp)>
  </Router>
</BrowserRouter>
`;

exports[`Sign Up View Component Should render without errors 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Connect(SignUp)
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />
</Router>
`;
